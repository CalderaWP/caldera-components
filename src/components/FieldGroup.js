import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup,FormControl,ControlLabel,HelpBlock } from 'react-bootstrap';
export function FieldGroup({ id, label, help, ...props }) {
	return (
		<FormGroup controlId={id}>
			<ControlLabel
				controlId={id}
			>
				{label}
			</ControlLabel>
			<FormControl {...props} />
			{help && <HelpBlock>{help}</HelpBlock>}
		</FormGroup>
	);
}

FieldGroup.propTypes = {
	id: PropTypes.string.isRequired,
	props: PropTypes.object,
	help: PropTypes.string,
	label: PropTypes.string.isRequired
};