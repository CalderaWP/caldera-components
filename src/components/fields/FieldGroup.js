import React from 'react';
import classNames from 'classnames';
import {addOrRemoveFromArray, ariaDescribedbyAttr} from './util';
import {
	fieldGroupPropTypes,
} from './propTypes';

import {FieldInner} from './FieldInner';
import {RenderGroup} from '../RenderGroup';
import {Message} from './messages/Message';
import {messageObjectFactory} from './messages/messageObjectFactory';

/**
 * Represents one configField -- wrapper, label and input.
 *
 * @param props
 * @return {*}
 * @constructor
 */
export const FieldGroup = (props) => {

	/**
	 * Creates the id attribute
	 * @return {String}
	 */
	function idAttrForHelpElement() {
		return ariaDescribedbyAttr(props.id, props.help);
	}

	function checkboxChangeHandler(option,values) {
		if (!Array.isArray(values)) {
			if ('string' === typeof values || 'number' === typeof  values) {
				values = [values];
			} else {
				values = [];

			}
		}
		values = addOrRemoveFromArray(option.value, values);
		props.onValueChange(values);
	}

	/**
	 * Based on type, render the field.
	 *
	 * @return {*}
	 */
	function fieldInner(fieldProps){
		const message = 'object' === typeof  fieldProps.message
			? messageObjectFactory(fieldProps.message)
			: messageObjectFactory({message:null, error: false });
		function getFieldClassName(conditionalClassNames) {
			return classNames(
				{
					required: fieldProps.isRequired,
					...conditionalClassNames
				}
			);
		}

		/**
		 * Is this option the checked value?
		 * @param {object}option
		 * @param {String|number|Array} currentValue
		 * @return {boolean}
		 */
		function isCheckedOption( option, currentValue){
			if( Array.isArray( currentValue ) ){
				return currentValue.includes(option.value);
			}
			return option.value === currentValue;
		}

		//Fieldsets are rendered recursively.
		if( 'fieldset' === fieldProps.type ){
			return (
				<fieldset
					id={props.id}
				>
					<legend>
						{props.label}
					</legend>
					{props.options.map(option =>  {


						//Call this same function, as a regular checkbox
						return (
							<React.Fragment
								key={ `${props.id}-${option.value}` }
							>
								{fieldInner({
									type: 'input',
									inputType: 'checkbox',
									id: `${props.id}-${option.value}`,
									value: isCheckedOption(option,props.value),
									fieldClassName: getFieldClassName({
										'fieldset-checkbox' : true,
									}),
									label: option.label,
									disabled:props.disabled,
									onValueChange: () => {
										checkboxChangeHandler(option,props.values);
									}
								})}
							</React.Fragment>
						);
					})}
				</fieldset>
			);
		}else if( 'hidden' === fieldProps.inputType ){
			return (
				FieldInner(
					{
						type: props.type,
						id: props.id,
						fieldClassName: getFieldClassName(),
						value:fieldProps.value,
						onValueChange:fieldProps.onValueChange,
						inputType: fieldProps.inputType,
						disabled:props.disabled,
					}
				)
			);
		}else{
			return (
				<div
					className={RenderGroup.classNames.fieldWrapper}
				>
					<label
						htmlFor={fieldProps.id}
					>
						{fieldProps.label}
					</label>
					<Message
						message={message}
					/>
					{FieldInner(
						{
							type: fieldProps.type,
							id: fieldProps.id,
							fieldClassName: getFieldClassName({
								'block-input': fieldProps.isBlockInput
							}),
							help:fieldProps.help,
							value:fieldProps.value,
							onValueChange:fieldProps.onValueChange,
							inputType: fieldProps.inputType,
							options: fieldProps.options,
							disabled:props.disabled,
						}
					)}
					{fieldProps.help &&
					<p
						id={`${idAttrForHelpElement()}`}
						className={'description'}
					>
						{fieldProps.help}
					</p>
					}
				</div>
			);
		}

	}

	/**
	 * Render field group
	 */
	return fieldInner(props);
};




FieldGroup.propTypes = fieldGroupPropTypes;

FieldGroup.defaultProps = {
	isBlockInput: true,
	isRequired: false,
	help: ''
};