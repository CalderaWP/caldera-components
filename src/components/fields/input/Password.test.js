import renderer from 'react-test-renderer';
import React from 'react';
import {shallow} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {Password} from './Password';
import {FieldGroup} from "../FieldGroup";

Enzyme.configure({adapter: new Adapter()});

describe('Input component', () => {
	it('Passes its props', () => {
		const component = renderer.create(
			<Password
				id={'bags'}
				fieldClassName={'foo'}
				onValueChange={() => {
				}}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	describe('Password component rendering', () => {

		it( 'passes value to change handler', () => {
			let receivedValue = '';
			const handler = (newValue) => {
				receivedValue = newValue;
			};
			const wrapper = shallow(
				<Password
					id={'password2'}
					fieldClassName={'password'}
					onValueChange={handler}
					value={receivedValue}
				/>
			);
			wrapper.find('#password2').simulate('change', { target: { value: '12345' } });
			expect(receivedValue).toBe('12345');
			//expect(wrapper.find('input').prop('type')).toBe('password');
		});

		it( 'Renders in FieldGroup', () => {

			const wrapper = shallow(
				<FieldGroup
					id={'password3'}
					label={'password'}
					type={'password'}
					innertype={'password'}
					isRequired={true}
					value={'12345'}
					onValueChange={() => {

					}}
					help={'aaa'}

				/>
			)
			expect(wrapper.find('#password3').prop('type')).toBe('password');
		});

	});


});



