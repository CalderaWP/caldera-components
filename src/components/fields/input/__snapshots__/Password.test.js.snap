// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Passes its props 1`] = `
<input
  aria-describedby=""
  className="foo field-config"
  disabled={undefined}
  id="bags"
  onBlur={undefined}
  onChange={[Function]}
  onFocus={undefined}
  required={false}
  type="password"
  value={undefined}
/>
`;
